### ISSUE TITLE
Considering simplifying `clip` and other APIs

### ISSUE THREAD

**Issue by joshgoebel** [nesbox/TIC-80#1770]
As mentioned in  #1259 the code to do some simply things at a low-level gets kind of ugly... for example to call `clip()` in WASM requires we pass all 4 arguments... if we use `-1` to say "default" we're left with this.

```wat
i32.const -1
i32.const -1
i32.const -1
i32.const -1
call $clip
```

Lately we've added a ton of new API which seems to me unnecessary (but convenient)... `peek2`, `peek1`, `poke2`, `poke1`, etc... if there room for a bit more unnecessary, but convenient (and perhaps more readable) API?

For example taking `clip` as an example case... instead of:

```js
clip(10,10,50,50) // setup clipping
clip()            // remove clipping
```

Why not:

```js
clip(10,10,50,50) // setup clipping
noclip()          // remove clipping
```

To me that reads nicer (more intentional) and at a lower-level of course we'd now have two functions so one would require no arguments so removing clipping is now a single instruction (vs 5):

```wat
call $noclip
```

Thoughts?

**Comment by joshgoebel** [comment]
The existing API could be left in place (for backwards compatibility) and `clip()` would internally just call `noclip()`... with the WASM API being an edge case (since it's new)... it could *perhaps* only support the newer API.

**Comment by paul59** [comment]
**I know next to nothing of things under the hood** but there are other inconsistencies that make 'the same' API call from the user perspective that would logically need attention if you go down the noclip() route wholeheartedly:

music(0) play, music() stop .... so nomusic()?

pix(1,1,1) write, pix(1,1) read... so getpixel()? We have fget() and fset() so why not?

Not exactly the same but: 

sfx(0) play, sfx(-1) stop... so stopsfx(0) or stopchannel(0)?

There might be others that don't come to mind at the minute!

In other words, introducing noclip(), for example, without addressing the others would just add another inconsistency - at least a parameterless call to clip() matches music() with regard to 'cancelling' something :D

Personally I'm easy either way, go with what we have or introduce noclip(), nomusic() etc

**Comment by joshgoebel** [comment]
I don't see the need for _everything_ you suggested... the real issue (for WASM at least) is the function with absurdly long arguments lists... so `pix` isn't a big problem.  I do think the consistency argument would apply to `pmem` though, that perhaps we should have a getter/setter combo... so if we want a slightly more thorough proposal:

Before:

-`music()` stop music
- `sfx(-1)` stop sound
- `sfx(-1,nil,nill,channel) stop sound on channel
- `pmem(addy)` read persistent ram
- `pmem(addy, value)` write persistent ram

After (ideas):

- `nomusic()` stop music
- `nosfx()` stop sound
- `nosfx(channel)` stop sound on channel
- `psget(addy)` read persistent ram
- `psset(addy, value)` write persistent ram

`no` is shorter but I could be easily persuaded to `stop` or adding an underscore in-between, etc... though that might get into naming consistencies between languages... `no_music` in Lua vs `noMusic` in JS... (not that we 100% have to follow each languages case conventions).

That still leaves `spr`, `font`, `print`, `map`, but those don't really have any easy simplifications off the top of my head...

**Comment by joshgoebel** [comment]
With regard to WASM this `-1` as default is still problematic in a few "differing types" (non-integers) places... C is far more "mushy" with types but other safer languages are more precise... so in C you can just throw a boolean in place of and `int` and it'll mostly work... but in Zig you'll get a compile time error and need to cast, this makes reusing some API's difficult like `btn`.

Lua:

```lua
btn() // returns a number
btn(1) // return a boolean
```

So our WASM API is currently `btn(int32) -> int32`, but this makes it harder to work with the result... you have to write code like:

```zig
if (btn(0) != 0) // because result is an `integer`
// instead of the much nicer
if (btn(0))  // result is a `bool`
```

This also bites us with boolean arguments to functions like print, where we really want to write this:

```zig
_ = tic.print(&buff, 1, 0, 11, false, 1, false);
```

Both `fixed` and `smallfont` are false... but then makes it impossible to write the following (to use defaults):

```zig
_ = tic.print(&buff, 1, 0, 11, -1, -1, -1);
```

Because `-1` is of type integer and does not match type `boolean`... if we just make the fields all `int32` then we can't pass booleans and have to use 0 and 1 which is uglier and harder to read.  Although perhaps this is solved for print by wrapping all the API arguments a struct... (so the actual WASM level API would still be all ints, just the higher level APIs could expose bools, etc...) - though this has some implications for documentation, etc...  

We probably first should have the discussion of whether we want to "own" the per-language bindings for our WASM layer or not... opening a new issue for that now.

**Comment by nesbox** [comment]
Let's not litter the current API and add helpers like `noclip ()` to the WASM libraries for each language as you mentioned in #1771

**Comment by joshgoebel** [comment]
Ok, I will close this for now.  Secretly I hope WASM becomes super popular and those richer APIs create a little back pressure on us to improve the main API a bit. 😉  A boy can dream.
